render/allocator.c

```diff
@@ -9,6 +18,49 @@ void wlr_allocator_init(struct wlr_allocator *alloc,
 	wl_signal_init(&alloc->events.destroy);
 }
 
+struct wlr_allocator *wlr_allocator_autocreate(struct wlr_backend *backend,
+		struct wlr_renderer *renderer) {
+	wlr_log(WLR_DEBUG, "Picking the appropriate allocator");
+	const uint32_t backend_caps = backend_get_buffer_caps(backend);
+	const uint32_t renderer_caps = renderer_get_render_buffer_caps(renderer);
+	const int drm_fd = wlr_backend_get_drm_fd(backend);
+
+	struct wlr_allocator *alloc = NULL;
+	const uint32_t gbm_caps = WLR_BUFFER_CAP_DMABUF;
+	if ((backend_caps & gbm_caps) && (renderer_caps & gbm_caps)
+			&& (drm_fd != -1)) {
+		wlr_log(WLR_DEBUG, "Trying to create gbm allocator");
+		const int fd = fcntl(drm_fd, F_DUPFD_CLOEXEC, 0);
+		if (fd < 0) {
+			wlr_log(WLR_ERROR, "fcntl(F_DUPFD_CLOEXEC) failed");
+		} else if ((alloc = wlr_gbm_allocator_create(fd)) != NULL) {
+			return alloc;
+		}
+		wlr_log(WLR_DEBUG, "Failed to create gbm allocator");
+	}
+
+	const uint32_t shm_caps = WLR_BUFFER_CAP_SHM | WLR_BUFFER_CAP_DATA_PTR;
+	if ((backend_caps & shm_caps) && (renderer_caps & shm_caps)) {
+		wlr_log(WLR_DEBUG, "Trying to create shm allocator");
+		if ((alloc = wlr_shm_allocator_create()) != NULL) {
+			return alloc;
+		}
+		wlr_log(WLR_DEBUG, "Failed to create shm allocator");
+	}
+
+	/* TODO: drm_dumb allocator
```

Comment:
Maybe remove this commented code for now, we don't have the DRM dumb allocator yet.